File: web/src/api/service/service.ts
import type {
  ServiceBasicUpdateReq,
  ServiceComponentDetailResp,
  ServiceComponentResp,
  ServiceComponentOverlayUpdateReq,
  ServiceDeployReq,
  ServiceDeployResp,
  ServiceEnvUpdateReq,
  ServicePaginatedResp,
  ServicePreviewReq,
  ServicePreviewResp,
  ServiceResp,
  ServiceCreateReq,
} from '@/gen/proto/orbit/v1/service/service';
import request from '@/utils/request';

export const serviceApi = {
  list(
    projectId: string,
    params?: {
      page?: number;
      per_page?: number;
      application_id?: string;
      status?: string;
      search?: string;
    }
  ): Promise<ServicePaginatedResp> {
    return request.get('/api/service', { params: { project_id: projectId, ...params } });
  },

  get(projectId: string, id: string): Promise<ServiceResp> {
    return request.get(`/api/service/${id}`, { params: { project_id: projectId } });
  },

  create(projectId: string, payload: ServiceCreateReq): Promise<ServiceResp> {
    return request.post('/api/service', payload, { params: { project_id: projectId } });
  },

  remove(projectId: string, id: string): Promise<void> {
    return request.delete(`/api/service/${id}`, { params: { project_id: projectId } });
  },

  getComponent(
    projectId: string,
    serviceId: string,
    componentId: string
  ): Promise<ServiceComponentDetailResp> {
    return request.get(`/api/service/${serviceId}/component/${componentId}`, {
      params: { project_id: projectId },
    });
  },

  updateComponent(
    projectId: string,
    serviceId: string,
    componentId: string,
    payload: ServiceComponentOverlayUpdateReq
  ): Promise<ServiceComponentResp> {
    return request.put(`/api/service/${serviceId}/component/${componentId}`, payload, {
      params: { project_id: projectId },
    });
  },

  updateBasic(projectId: string, id: string, payload: ServiceBasicUpdateReq): Promise<ServiceResp> {
    return request.put(`/api/service/${id}/basic`, payload, { params: { project_id: projectId } });
  },

  updateEnv(projectId: string, id: string, payload: ServiceEnvUpdateReq): Promise<ServiceResp> {
    return request.put(`/api/service/${id}/env`, payload, { params: { project_id: projectId } });
  },

  preview(projectId: string, id: string, data: ServicePreviewReq): Promise<ServicePreviewResp> {
    return request.post(`/api/service/${id}/preview`, data, { params: { project_id: projectId } });
  },

  deploy(projectId: string, id: string, payload: ServiceDeployReq): Promise<ServiceDeployResp> {
    return request.post(`/api/service/${id}/deploy`, payload, {
      params: { project_id: projectId },
    });
  },
};

File: internal/api/http/handler/service/service.go
package servicehandler

import (
	"net/http"

	"github.com/leoninew/pomelo-orbit/internal/api/http/transport"

	"github.com/gin-gonic/gin"
	servicedto "github.com/leoninew/pomelo-orbit/internal/application/service/dto"
	servicev1 "github.com/leoninew/pomelo-orbit/internal/gen/proto/orbit/v1/service"
)

func (h Handler) ListServices(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	page := transport.QueryInt(c.Request.URL.Query().Get("page"), 1)
	perPage := transport.QueryInt(c.Request.URL.Query().Get("per_page"), 10)
	items, err := h.service.ListServices(c.Request.Context(), current.Id, servicedto.ServiceListInput{
		ProjectId:     c.Request.URL.Query().Get("project_id"),
		ApplicationId: c.Request.URL.Query().Get("application_id"),
		Status:        c.Request.URL.Query().Get("status"),
		Search:        c.Request.URL.Query().Get("search"),
		Page:          page,
		PerPage:       perPage,
	})
	if err != nil {
		transport.WriteError(c, err)
		return
	}
	resp := serviceViewResponses(items.Items)
	transport.WriteProtoJSON(c, http.StatusOK, &servicev1.ServicePaginatedResp{
		Items:   transport.Ptrs(resp),
		Total:   int32(items.Total),
		Page:    int32(items.Page),
		PerPage: int32(items.PerPage),
		Pages:   int32(transport.PageCount(items.Total, items.PerPage)),
	})
}

func (h Handler) GetService(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	view, err := h.service.GetService(c.Request.Context(), current.Id, c.Param("service_id"))
	if err != nil {
		transport.WriteError(c, err)
		return
	}
	resp := serviceViewResponse(view)
	transport.WriteProtoJSON(c, http.StatusOK, &resp)
}

func (h Handler) CreateService(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req servicev1.ServiceCreateReq
	if err := transport.DecodeJSON(c, &req); err != nil {
		transport.WriteStatusError(c, http.StatusBadRequest, "Invalid JSON body")
		return
	}
	view, err := h.service.CreateService(c.Request.Context(), current.Id, servicedto.ServiceCreateInput{
		ApplicationId: req.ApplicationId, VersionId: req.VersionId, InstanceKey: req.InstanceKey, Code: req.Code,
	})
	if err != nil {
		transport.WriteError(c, err)
		return
	}
	resp := serviceViewResponse(view)
	transport.WriteProtoJSON(c, http.StatusCreated, &resp)
}

func (h Handler) DeleteService(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	if err := h.service.DeleteService(c.Request.Context(), current.Id, c.Param("service_id")); err != nil {
		transport.WriteError(c, err)
		return
	}
	c.Status(http.StatusNoContent)
}

func (h Handler) GetServiceComponent(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	detail, err := h.service.GetServiceComponent(c.Request.Context(), current.Id, c.Param("service_id"), c.Param("component_id"))
	if err != nil {
		transport.WriteError(c, err)
		return
	}
	resp := serviceComponentDetailResponse(detail)
	transport.WriteProtoJSON(c, http.StatusOK, &resp)
}

func (h Handler) UpdateServiceComponentOverlay(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req servicev1.ServiceComponentOverlayUpdateReq
	if err := transport.DecodeJSON(c, &req); err != nil {
		transport.WriteStatusError(c, http.StatusBadRequest, "Invalid JSON body")
		return
	}
	component, err := h.service.UpdateServiceComponentOverlay(c.Request.Context(), current.Id, c.Param("service_id"), c.Param("component_id"), serviceComponentOverlayInput(&req))
	if err != nil {
		transport.WriteError(c, err)
		return
	}
	resp := serviceComponentResponse(component)
	transport.WriteProtoJSON(c, http.StatusOK, &resp)
}

func (h Handler) UpdateServiceBasic(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req servicev1.ServiceBasicUpdateReq
	if err := transport.DecodeJSON(c, &req); err != nil {
		transport.WriteStatusError(c, http.StatusBadRequest, "Invalid JSON body")
		return
	}
	view, err := h.service.UpdateServiceBasic(c.Request.Context(), current.Id, c.Param("service_id"), servicedto.ServiceBasicUpdateInput{
		VersionId: req.VersionId, InstanceKey: req.InstanceKey,
	})
	if err != nil {
		transport.WriteError(c, err)
		return
	}
	resp := serviceViewResponse(view)
	transport.WriteProtoJSON(c, http.StatusOK, &resp)
}

func (h Handler) UpdateServiceEnv(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req servicev1.ServiceEnvUpdateReq
	if err := transport.DecodeJSON(c, &req); err != nil {
		transport.WriteStatusError(c, http.StatusBadRequest, "Invalid JSON body")
		return
	}
	view, err := h.service.UpdateServiceEnv(c.Request.Context(), current.Id, c.Param("service_id"), serviceEnvUpdateInput(&req))
	if err != nil {
		transport.WriteError(c, err)
		return
	}
	resp := serviceViewResponse(view)
	transport.WriteProtoJSON(c, http.StatusOK, &resp)
}

// ListApplicationServices uses the service domain's authorization and query path.
func (h Handler) ListApplicationServices(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	items, err := h.service.ListServiceViewsByApplication(c.Request.Context(), current.Id, c.Param("app_id"))
	if err != nil {
		transport.WriteError(c, err)
		return
	}
	resp := serviceViewResponses(items)
	transport.WriteProtoJSON(c, http.StatusOK, &servicev1.ServiceListResp{Items: transport.Ptrs(resp)})
}
